cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -361,6 +374,101 @@ public boolean apply(Id.Program programId) {
   }
 
   /**
+   * Upgrade from the previous version of CDAP to the current version. In 3.2, this means adding artifacts for all
+   * existing applications and updating their specs to include the artifact id.
+   */
+  public void upgrade(boolean continueOnFailure) throws Exception {
+    File tmpDir = new File(configuration.get(Constants.CFG_LOCAL_DATA_DIR),
+      configuration.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile();
+    if (!tmpDir.exists()) {
+      if (!tmpDir.mkdirs()) {
+        throw new IOException("Unable to create tmp dir " + tmpDir.getAbsolutePath());
+      }
+    }
+
+    for (NamespaceMeta namespaceMeta : store.listNamespaces()) {
+      Id.Namespace namespaceId = Id.Namespace.from(namespaceMeta.getName());
+      for (ApplicationSpecification appSpec : store.getAllApplications(namespaceId)) {
+        Id.Application appId = Id.Application.from(namespaceId, appSpec.getName());
+
+        if (appSpec.getArtifactId() != null) {
+          // this is possible if upgrade failed midway through. It means this app was already updated so continue on.
+          continue;
+        }
+
+        Location appJarLocation = store.getApplicationArchiveLocation(appId);
+        if (appJarLocation == null) {
+          LOG.error(String.format("Unable to get location of jar for app '%s' in namespace '%s'. " +
+              "You will need to re-deploy the app after upgrade.",
+            appId.getNamespaceId(), appId.getId()));
+          continue;
+        }
+
+        // copy jar to local file
+        File tmpFile = File.createTempFile("tmpApp", ".jar", tmpDir);
+        Files.copy(Locations.newInputSupplier(appJarLocation), tmpFile);
+
+        // use app name as artifact name
+        String version = appSpec.getVersion();
+        if (version == null || version.isEmpty()) {
+          // version derived from manifest.
+          JarFile jarFile = new JarFile(tmpFile);
+          Manifest manifest = jarFile.getManifest();
+          if (manifest != null && manifest.getMainAttributes() != null) {
+            version = manifest.getMainAttributes().getValue(ManifestFields.BUNDLE_VERSION);
+          }
+          // If version couldn't be found through manifest, default it to 1.0.0
+          if (version == null || version.isEmpty()) {
+            version = "1.0.0";
+          }
+        }
+
+        ArtifactDetail artifactDetail;
+        Id.Artifact artifactId = Id.Artifact.from(namespaceId, appId.getId(), version);
+        try {
+          artifactDetail = artifactRepository.addArtifact(artifactId, tmpFile);
+        } catch (WriteConflictException e) {
```

Comment:
the scenario you mention is captured in the catch below for ArtifactAlreadyExists. In that scenario, it just looks up the info from the store. It is only possible if the upgrade tool ran before, the artifact was added, but the app metadata failed to update.  It is idempotent